pkg/sdn/plugin/sdn-cni-plugin/openshift-sdn.go

```diff
@@ -0,0 +1,611 @@
+	if err != nil {
+		return "", "", fmt.Errorf("failed to parse pod bandwidth: %v", err)
+	}
+	var ingressStr, egressStr string
+	if ingress != nil {
+		ingressStr = fmt.Sprintf("%d", ingress.Value())
+	}
+	if egress != nil {
+		egressStr = fmt.Sprintf("%d", egress.Value())
+	}
+	return ingressStr, egressStr, nil
+}
+
+func wantsMacvlan(pi *podInfo) (bool, error) {
+	val, found := pi.Annotations[AssignMacVlanAnnotation]
+	if !found || val != "true" {
+		return false, nil
+	}
+	if pi.Privileged {
+		return true, nil
+	}
+	return false, fmt.Errorf("pod has %q annotation but is not privileged", AssignMacVlanAnnotation)
+}
+
+func loadNetConf(bytes []byte) (*api.CNINetConfig, error) {
+	n := &api.CNINetConfig{}
+	if err := json.Unmarshal(bytes, n); err != nil {
+		return nil, fmt.Errorf("failed to load netconf: %v", err)
+	}
+	return n, nil
+}
+
+func getClients(masterKubeConfig string) (*oclient.Client, *kclient.Client, error) {
+	// Create OpenShift and Kubernetes clients to read the ClusterNetwork and PodSpec
+	overrides := &configapi.ClientConnectionOverrides{}
+	configapi.SetProtobufClientDefaults(overrides)
+
+	originClient, _, err := configapi.GetOpenShiftClient(masterKubeConfig, overrides)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	kubeClient, _, err := configapi.GetKubeClient(masterKubeConfig, overrides)
+	if err != nil {
+		return nil, nil, err
+	}
+	return originClient, kubeClient, nil
+}
+
+func getPodConfig(args *skel.CmdArgs) (*PodConfig, error) {
+	n, err := loadNetConf(args.StdinData)
+	if err != nil {
+		return nil, err
+	}
+
+	originClient, kubeClient, err := getClients(n.MasterKubeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// Grab the node's network configuration
+	cn, err := originClient.ClusterNetwork().Get(osapi.ClusterNetworkDefault)
+	if err != nil {
+		return nil, err
+	} else if _, _, err := net.ParseCIDR(cn.Network); err != nil {
+		return nil, err
+	}
+
+	hostSubnet, err := originClient.HostSubnets().Get(n.NodeName)
+	if err != nil {
+		return nil, err
+	} else if _, _, err := net.ParseCIDR(hostSubnet.Subnet); err != nil {
+		return nil, err
+	}
+
+	podInfo, err := readPodInfo(originClient, kubeClient, args.Args, cn.PluginName)
+	if err != nil {
+		return nil, err
+	}
+
+	ingress, egress, err := getBandwidth(podInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	macvlan, err := wantsMacvlan(podInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	return &PodConfig{
+		masterKubeConfig: n.MasterKubeConfig,
+		clusterNetwork:   cn.Network,
+		nodeNetwork:      hostSubnet.Subnet,
+		mtu:              n.MTU,
+		vnid:             strconv.FormatUint(uint64(podInfo.Vnid), 10),
+		ingressBandwidth: ingress,
+		egressBandwidth:  egress,
+		wantMacvlan:      macvlan,
+		podStatusPath:    podInfo.podStatusPath,
+	}, nil
+}
+
+// Returns host veth name, container veth MAC, and pod IP
+func getVethInfo(netns, containerIfname string) (netlink.Link, string, string, error) {
```

Comment:
the comment says it returns "host veth name", but it actually returns the whole netlink.Link, even though all its callers only want the name